Make DPCL compression more aggressive to avoid exceeding pubsub size limit

milan-zededa opened the pull request:
# Description

Refactored the DPCL compression logic to more aggressively remove obsolete DevicePortConfig (DPC) entries, preventing the DevicePortConfigList (DPCL) from exceeding the pubsub message size limit (64 KB).

Previously, DPCL compression would only occur if the latest DPC provided working controller connectivity. However, if the device was operating in a fallback state (i.e., using an older config) and the user attempted to fix networking through multiple unsuccessful configuration changes, this could lead to DPCL growing beyond the pubsub size limit. Since DPCL is persisted, this situation could result in a pubsub panic and an endless reboot loop.

The new compression logic is always applied (unless verification is in progress) and retains only:
- The latest DPC (highest priority, index 0)
- The most recent DPC with working connectivity (if any)
- The currently used DPC (as indicated by `CurrentIndex`)
- A manually injected `manual` DPC (via TUI)
- A `lastresort` DPC (if `network.fallback.any.eth` is enabled)

All other DPCs are removed. For discarded DPCs injected via the installer, their hash value is saved using `SaveShaInFile` to prevent re-injection on reboot (no changes there from this patch).

The `CurrentIndex` is adjusted as needed to reflect the retained entry. Compression is skipped if DPC verification is currently ongoing.

## How to test and validate this PR

1. Onboard an EVE device.
2. From the controller, edit the network configuration for the device's port(s):
  - Set the network type to static.
  - Intentionally configure incorrect IP settings (e.g., a subnet that does not match the actual network the device is connected to).
3. After EVE receives and tests this new config:
  - It will fail to connect to the controller using the bad config.
  -  EVE will fall back to the previous working config and publish an error for the failed attempt.
4. On the device, verify that both the old and new configs are present, and that the fallback (previous) config is active:
```
jq '.CurrentIndex' /persist/status/nim/DevicePortConfigList/global.json
# Expected output: 1
# Note: index 0 is for the latest config, index 1 is the previous config.

jq '.PortConfigList | length' /persist/status/nim/DevicePortConfigList/global.json
# Expected output: 2 (two network configs persisted)
```
5. Repeat step 2 several times, each time pushing a different incorrect static IP configuration.
 
Previously, EVE would keep accumulating these incorrect configs in `/persist/status/nim/DevicePortConfigList/global.json`.
Eventually, the file would exceed the 64KB limit, causing PubSub to panic due to message size constraints and device entering an infinite reboot loop. With this patch, EVE should now keep only the latest DPC and the initial good network config.

To verify the fix, run the same commands again on the device:
```
jq '.CurrentIndex' /persist/status/nim/DevicePortConfigList/global.json
# Again Expected: 1

jq '.PortConfigList | length' /persist/status/nim/DevicePortConfigList/global.json
# Again Expected: 2
```
Confirm that even after multiple failed config pushes, only two entries are retained and the file size remains well below the 64KB limit.

## Changelog notes

Make DPCL compression more aggressive to avoid exceeding pubsub size limit

## PR Backports

This is quite serious bug, so it should go to every active LTS:

- 14.5-stable
- 13.4-stable

## Checklist

- [x] I've provided a proper description
- [ ] I've added the proper documentation
- [x] I've tested my PR on amd64 device
- [ ] I've tested my PR on arm64 device
- [x] I've written the test verification instructions
- [x] I've set the proper labels to this PR
- [x] I've checked the boxes above, or I've provided a good reason why I didn't check them.

eriknordmark reviewed (commented):
The current index can be -1 if there are failures. Should we make sure to keep at least one (with highest time priority) for zedagent (or for each source in general - can't there be multiple manual ones?)

milan-zededa commented:
> The current index can be -1 if there are failures. Should we make sure to keep at least one (with highest time priority) for zedagent (or for each source in general - can't there be multiple manual ones?)

That makes sense -- I've updated the PR to always retain the latest config from the controller (key "zedagent").
However, I would avoid keeping bootstrap or override configs indefinitely. Re-testing those long after the device has been onboarded is likely not meaningful or useful.

milan-zededa commented:
> > The current index can be -1 if there are failures. Should we make sure to keep at least one (with highest time priority) for zedagent (or for each source in general - can't there be multiple manual ones?)
> 
> Should this `-1` index have been described in the doc then? I see that only zero is described as a special value for the current config:
> 
> https://github.com/lf-edge/eve/blob/b61d6fb6ebae9959d82f04bbbf1c50d6ebf1cb78/docs/DEVICE-CONNECTIVITY.md?plain=1#L639
> and
> 
> https://github.com/lf-edge/eve/blob/b61d6fb6ebae9959d82f04bbbf1c50d6ebf1cb78/docs/DEVICE-CONNECTIVITY.md?plain=1#L426

Added explanation for the `-1` value into the documentation.

eriknordmark commented:
> > The current index can be -1 if there are failures. Should we make sure to keep at least one (with highest time priority) for zedagent (or for each source in general - can't there be multiple manual ones?)
> 
> That makes sense -- I've updated the PR to always retain the latest config from the controller (key "zedagent").

Great.

> However, I would avoid keeping bootstrap or override configs indefinitely. Re-testing those long after the device has been onboarded is likely not meaningful or useful (i.e. we will not keep at least one from each source).

Old/current code will prune those once the highest priority DPC is working.
I don't know what behavior we'd want before then when we potentially have some fallback, bootstrap, manual and not yet a zedagent DPC. Based on the textual description in the PR you will keep a/one fallback and one manual, but not the bootstrap/override one. That seems a bit odd to me.

A more uniform way would be to keep the most recent for each key (and once we have working DPC those will be pruned as in the old/current code).

milan-zededa commented:
> > > The current index can be -1 if there are failures. Should we make sure to keep at least one (with highest time priority) for zedagent (or for each source in general - can't there be multiple manual ones?)
> > 
> > 
> > That makes sense -- I've updated the PR to always retain the latest config from the controller (key "zedagent").
> 
> Great.
> 
> > However, I would avoid keeping bootstrap or override configs indefinitely. Re-testing those long after the device has been onboarded is likely not meaningful or useful (i.e. we will not keep at least one from each source).
> 
> Old/current code will prune those once the highest priority DPC is working. I don't know what behavior we'd want before then when we potentially have some fallback, bootstrap, manual and not yet a zedagent DPC. Based on the textual description in the PR you will keep a/one fallback and one manual, but not the bootstrap/override one. That seems a bit odd to me.
> 
> A more uniform way would be to keep the most recent for each key (and once we have working DPC those will be pruned as in the old/current code).

I have added a new rule:
```
- If there is no DPC from the controller (key `zedagent`), or if none of
  the controller-provided DPCs have ever succeeded in a connectivity test,
  retain the most recent DPC from each source to ensure fallback options
  remain available
```

eriknordmark commented:
> I have added a new rule:
> 
> ```
> - If there is no DPC from the controller (key `zedagent`), or if none of
>   the controller-provided DPCs have ever succeeded in a connectivity test,
>   retain the most recent DPC from each source to ensure fallback options
>   remain available
> ```

A bit complex. but OK.
